[Android][Java] Remove deprecated Java finalize to unblock android chip-tool compilation.

Since Java 9, Finalize has been marked as deprecated, compiler are showing them as warning, our matter rules.gni is treating warning as error, so that local compilation fails.

We need update the below files to remove finalize.

2025-02-04 07:01:40 INFO    [1716/1730] ACTION //src/controller/java:android_chip_im__javac(//build/toolchain/android:android_arm64)
2025-02-04 07:01:40 INFO    FAILED: gen/src/controller/java/android_chip_im.classlist
2025-02-04 07:01:40 INFO    python3 ../../build/chip/java/javac_runner.py --classdir obj/src/controller/java/android_chip_im/classes --outfile gen/src/controller/java/android_chip_im.classlist --build-config gen/src/controller/java/android_chip_im.json -- -d obj/src/controller/java/android_chip_im/classes @gen/src/controller/java/android_chip_im.sources -Werror -Xlint:all -Xlint:deprecation -parameters -Xlint:-options -source 8 -target 8
2025-02-04 07:01:40 INFO    ../../src/controller/java/src/chip/devicecontroller/ExtendableInvokeCallbackJni.java:86: warning: [removal] finalize() in Object has been deprecated and marked for removal
2025-02-04 07:01:40 INFO      protected void finalize() throws Throwable {
2025-02-04 07:01:40 INFO                     ^
2025-02-04 07:01:40 INFO    ../../src/controller/java/src/chip/devicecontroller/ExtendableInvokeCallbackJni.java:87: warning: [removal] finalize() in Object has been deprecated and marked for removal
2025-02-04 07:01:40 INFO        super.finalize();
2025-02-04 07:01:40 INFO             ^
2025-02-04 07:01:40 INFO    error: warnings found and -Werror specified
2025-02-04 07:01:40 INFO    ../../src/controller/java/src/chip/devicecontroller/GetConnectedDeviceCallbackJni.java:40: warning: [removal] finalize() in Object has been deprecated and marked for removal
2025-02-04 07:01:40 INFO      protected void finalize() throws Throwable {
2025-02-04 07:01:40 INFO                     ^
2025-02-04 07:01:40 INFO    ../../src/controller/java/src/chip/devicecontroller/GetConnectedDeviceCallbackJni.java:41: warning: [removal] finalize() in Object has been deprecated and marked for removal
2025-02-04 07:01:40 INFO        super.finalize();
2025-02-04 07:01:40 INFO             ^
2025-02-04 07:01:40 INFO    ../../src/controller/java/src/chip/devicecontroller/InvokeCallbackJni.java:61: warning: [removal] finalize() in Object has been deprecated and marked for removal
2025-02-04 07:01:40 INFO      protected void finalize() throws Throwable {
2025-02-04 07:01:40 INFO                     ^
2025-02-04 07:01:40 INFO    ../../src/controller/java/src/chip/devicecontroller/InvokeCallbackJni.java:62: warning: [removal] finalize() in Object has been deprecated and marked for removal
2025-02-04 07:01:40 INFO        super.finalize();
2025-02-04 07:01:40 INFO             ^
2025-02-04 07:01:40 INFO    ../../src/controller/java/src/chip/devicecontroller/ReportCallbackJni.java:94: warning: [removal] finalize() in Object has been deprecated and marked for removal
2025-02-04 07:01:40 INFO      protected void finalize() throws Throwable {
2025-02-04 07:01:40 INFO                     ^
2025-02-04 07:01:40 INFO    ../../src/controller/java/src/chip/devicecontroller/ReportCallbackJni.java:95: warning: [removal] finalize() in Object has been deprecated and marked for removal
2025-02-04 07:01:40 INFO        super.finalize();
2025-02-04 07:01:40 INFO             ^
2025-02-04 07:01:40 INFO    ../../src/controller/java/src/chip/devicecontroller/WriteAttributesCallbackJni.java:67: warning: [removal] finalize() in Object has been deprecated and marked for removal
2025-02-04 07:01:40 INFO      protected void finalize() throws Throwable {
2025-02-04 07:01:40 INFO                     ^
2025-02-04 07:01:40 INFO    ../../src/controller/java/src/chip/devicecontroller/WriteAttributesCallbackJni.java:68: warning: [removal] finalize() in Object has been deprecated and marked for removal
2025-02-04 07:01:40 INFO        super.finalize();
2025-02-04 07:01:40 INFO             ^
2025-02-04 07:01:40 INFO    ../../src/controller/java/generated/java/chip/devicecontroller/ChipClusters.java:168: warning: [removal] finalize() in Object has been deprecated and marked for removal
2025-02-04 07:01:40 INFO        protected void finalize() throws Throwable {
2025-02-04 07:01:40 INFO                       ^
2025-02-04 07:01:40 INFO    ../../src/controller/java/generated/java/chip/devicecontroller/ChipClusters.java:169: warning: [removal] finalize() in Object has been deprecated and marked for removal
2025-02-04 07:01:40 INFO          super.finalize();
2025-02-04 07:01:40 INFO               ^
2025-02-04 07:01:40 INFO    1 error
2025-02-04 07:01:40 INFO    12 warnings
2025-02-04 07:01:40 INFO    Traceback (most recent call last):
2025-02-04 07:01:40 INFO      File "/usr/local/google/home/yunhanw/workspace/connectedhomeip/out/android-arm64-chip-tool/../../build/chip/java/javac_runner.py", line 146, in <module>
2025-02-04 07:01:40 INFO        sys.exit(main())
2025-02-04 07:01:40 INFO                 ^^^^^^
2025-02-04 07:01:40 INFO      File "/usr/local/google/home/yunhanw/workspace/connectedhomeip/out/android-arm64-chip-tool/../../build/chip/java/javac_runner.py", line 130, in main
2025-02-04 07:01:40 INFO        retcode = subprocess.check_call(java_args + args.rest)
2025-02-04 07:01:40 INFO                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-02-04 07:01:40 INFO      File "/usr/lib/python3.11/subprocess.py", line 413, in check_call
2025-02-04 07:01:40 INFO        raise CalledProcessError(retcode, cmd)
2025-02-04 07:01:40 INFO    subprocess.CalledProcessError: Command '['/usr/local/buildtools/java/jdk/bin/javac', '-classpath', 'lib/src/controller/java/libMatterTlv.jar:lib/src/controller/java/CHIPClusters.jar:lib/third_party/java_deps/jsr305-3.0.2.jar:lib/src/controller/java/android.jar:lib/src/controller/java/CHIPClusterID.jar', '-d', 'obj/src/controller/java/android_chip_im/classes', '@gen/src/controller/java/android_chip_im.sources', '-Werror', '-Xlint:all', '-Xlint:deprecation', '-parameters', '-Xlint:-options', '-source', '8', '-target', '8']' returned non-zero exit status 1.
2025-02-04 07:02:03 INFO    [1717/1730] ACTION //:matter_build_venv._install_3p_deps(//third_party/pigweed/repo/pw_build/python_toolchain:python)
2025-02-04 07:02:03 INFO    ninja: build stopped: subcommand failed.

--- a/scripts/py_matter_idl/matter_idl/generators/java/ChipClusters_java.jinja
+++ b/scripts/py_matter_idl/matter_idl/generators/java/ChipClusters_java.jinja
@@ -105,6 +105,7 @@ import chip.devicecontroller.model.NodeState;
 import chip.devicecontroller.model.Status;
 
 import javax.annotation.Nullable;
+import java.lang.ref.Cleaner;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
@@ -169,10 +170,19 @@ public class ChipClusters {
 
     private Optional<Long> timeoutMillis = Optional.empty();
 
+    private final Cleaner.Cleanable cleanable;
+
     public BaseChipCluster(long devicePtr, int endpointId, long clusterId) {
       this.devicePtr = devicePtr;
       this.endpointId = endpointId;
       this.clusterId = clusterId;
+
+      this.cleanable = Cleaner.create().register(this, () -> {
+         if (chipClusterPtr != 0) {
+           deleteCluster(chipClusterPtr);
+           chipClusterPtr = 0;
+         }
+      });
     }
 
     /**
@@ -241,15 +251,6 @@ public class ChipClusters {
 
     @Deprecated
     public void deleteCluster(long chipClusterPtr) {}
-    @SuppressWarnings("deprecation")
-    protected void finalize() throws Throwable {
-      super.finalize();
-
-      if (chipClusterPtr != 0) {
-        deleteCluster(chipClusterPtr);
-        chipClusterPtr = 0;
-      }
-    }
   }
 
   abstract static class ReportCallbackImpl implements ReportCallback, SubscriptionEstablishedCallback, ResubscriptionAttemptCallback {

--- a/scripts/py_matter_idl/matter_idl/generators/java/ChipClusters_java.jinja
+++ b/scripts/py_matter_idl/matter_idl/generators/java/ChipClusters_java.jinja
@@ -177,12 +177,16 @@ public class ChipClusters {
       this.endpointId = endpointId;
       this.clusterId = clusterId;
 
-      this.cleanable = Cleaner.create().register(this, () -> {
-         if (chipClusterPtr != 0) {
-           deleteCluster(chipClusterPtr);
-           chipClusterPtr = 0;
-         }
-      });
+      this.cleanable =
+        Cleaner.create()
+          .register(
+            this,
+            () -> {
+              if (chipClusterPtr != 0) {
+                deleteCluster(chipClusterPtr);
+                chipClusterPtr = 0;
+              }
+            });
     }
 
     /**

--- a/scripts/py_matter_idl/matter_idl/tests/outputs/several_clusters/java/ChipClusters.java
+++ b/scripts/py_matter_idl/matter_idl/tests/outputs/several_clusters/java/ChipClusters.java
@@ -100,12 +100,16 @@ public class ChipClusters {
       this.endpointId = endpointId;
       this.clusterId = clusterId;
 
-      this.cleanable = Cleaner.create().register(this, () -> {
-         if (chipClusterPtr != 0) {
-           deleteCluster(chipClusterPtr);
-           chipClusterPtr = 0;
-         }
-      });
+      this.cleanable =
+        Cleaner.create()
+          .register(
+            this,
+            () -> {
+              if (chipClusterPtr != 0) {
+                deleteCluster(chipClusterPtr);
+                chipClusterPtr = 0;
+              }
+            });
     }
 
     /**

--- a/scripts/py_matter_idl/matter_idl/tests/outputs/several_clusters/java/ChipClusters.java
+++ b/scripts/py_matter_idl/matter_idl/tests/outputs/several_clusters/java/ChipClusters.java
@@ -28,6 +28,7 @@ import chip.devicecontroller.model.NodeState;
 import chip.devicecontroller.model.Status;
 
 import javax.annotation.Nullable;
+import java.lang.ref.Cleaner;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
@@ -92,10 +93,19 @@ public class ChipClusters {
 
     private Optional<Long> timeoutMillis = Optional.empty();
 
+    private final Cleaner.Cleanable cleanable;
+
     public BaseChipCluster(long devicePtr, int endpointId, long clusterId) {
       this.devicePtr = devicePtr;
       this.endpointId = endpointId;
       this.clusterId = clusterId;
+
+      this.cleanable = Cleaner.create().register(this, () -> {
+         if (chipClusterPtr != 0) {
+           deleteCluster(chipClusterPtr);
+           chipClusterPtr = 0;
+         }
+      });
     }
 
     /**
@@ -164,15 +174,6 @@ public class ChipClusters {
 
     @Deprecated
     public void deleteCluster(long chipClusterPtr) {}
-    @SuppressWarnings("deprecation")
-    protected void finalize() throws Throwable {
-      super.finalize();
-
-      if (chipClusterPtr != 0) {
-        deleteCluster(chipClusterPtr);
-        chipClusterPtr = 0;
-      }
-    }
   }
 
   abstract static class ReportCallbackImpl implements ReportCallback, SubscriptionEstablishedCallback, ResubscriptionAttemptCallback {

--- a/src/controller/java/generated/java/chip/devicecontroller/ChipClusters.java
+++ b/src/controller/java/generated/java/chip/devicecontroller/ChipClusters.java
@@ -100,12 +100,16 @@ public class ChipClusters {
       this.endpointId = endpointId;
       this.clusterId = clusterId;
 
-      this.cleanable = Cleaner.create().register(this, () -> {
-         if (chipClusterPtr != 0) {
-           deleteCluster(chipClusterPtr);
-           chipClusterPtr = 0;
-         }
-      });
+      this.cleanable =
+        Cleaner.create()
+          .register(
+            this,
+            () -> {
+              if (chipClusterPtr != 0) {
+                deleteCluster(chipClusterPtr);
+                chipClusterPtr = 0;
+              }
+            });
     }
 
     /**

--- a/src/controller/java/generated/java/chip/devicecontroller/ChipClusters.java
+++ b/src/controller/java/generated/java/chip/devicecontroller/ChipClusters.java
@@ -28,6 +28,7 @@ import chip.devicecontroller.model.NodeState;
 import chip.devicecontroller.model.Status;
 
 import javax.annotation.Nullable;
+import java.lang.ref.Cleaner;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
@@ -92,10 +93,19 @@ public class ChipClusters {
 
     private Optional<Long> timeoutMillis = Optional.empty();
 
+    private final Cleaner.Cleanable cleanable;
+
     public BaseChipCluster(long devicePtr, int endpointId, long clusterId) {
       this.devicePtr = devicePtr;
       this.endpointId = endpointId;
       this.clusterId = clusterId;
+
+      this.cleanable = Cleaner.create().register(this, () -> {
+         if (chipClusterPtr != 0) {
+           deleteCluster(chipClusterPtr);
+           chipClusterPtr = 0;
+         }
+      });
     }
 
     /**
@@ -164,15 +174,6 @@ public class ChipClusters {
 
     @Deprecated
     public void deleteCluster(long chipClusterPtr) {}
-    @SuppressWarnings("deprecation")
-    protected void finalize() throws Throwable {
-      super.finalize();
-
-      if (chipClusterPtr != 0) {
-        deleteCluster(chipClusterPtr);
-        chipClusterPtr = 0;
-      }
-    }
   }
 
   abstract static class ReportCallbackImpl implements ReportCallback, SubscriptionEstablishedCallback, ResubscriptionAttemptCallback {

--- a/src/controller/java/src/chip/devicecontroller/BatchInvokeCallbackJni.java
+++ b/src/controller/java/src/chip/devicecontroller/BatchInvokeCallbackJni.java
@@ -19,17 +19,27 @@ package chip.devicecontroller;
 
 import chip.devicecontroller.model.InvokeResponseData;
 import chip.devicecontroller.model.NoInvokeResponseData;
+import java.lang.ref.Cleaner;
 import java.util.Optional;
 import javax.annotation.Nullable;
 
 /** JNI wrapper callback class for {@link InvokeCallback}. */
-public final class ExtendableInvokeCallbackJni {
-  private final ExtendableInvokeCallback wrappedExtendableInvokeCallback;
+public final class BatchInvokeCallbackJni {
+  private final BatchInvokeCallbackJni wrappedBatchInvokeCallback;
   private long callbackHandle;
 
-  public ExtendableInvokeCallbackJni(ExtendableInvokeCallback wrappedExtendableInvokeCallback) {
-    this.wrappedExtendableInvokeCallback = wrappedExtendableInvokeCallback;
+  private final Cleaner.Cleanable cleanable;
+
+  public BatchInvokeCallbackJni(BatchInvokeCallback wrappedBatchInvokeCallback) {
+    this.wrappedBatchInvokeCallback = wrappedBatchInvokeCallback;
     this.callbackHandle = newCallback();
+
+    this.cleanable = Cleaner.create().register(this, () -> {
+      if (chipClusterPtr != 0) {
+        deleteCluster(chipClusterPtr);
+        chipClusterPtr = 0;
+      }
+    });
   }
 
   long getCallbackHandle() {
@@ -41,7 +51,7 @@ public final class ExtendableInvokeCallbackJni {
   private native void deleteCallback(long callbackHandle);
 
   private void onError(Exception e) {
-    wrappedExtendableInvokeCallback.onError(e);
+    wrappedBatchInvokeCallback.onError(e);
   }
 
   private void onResponse(
@@ -74,21 +84,10 @@ public final class ExtendableInvokeCallbackJni {
   }
 
   private void onNoResponse(int commandRef) {
-    wrappedExtendableInvokeCallback.onNoResponse(NoInvokeResponseData.newInstance(commandRef));
+    wrappedBatchInvokeCallback.onNoResponse(NoInvokeResponseData.newInstance(commandRef));
   }
 
   private void onDone() {
-    wrappedExtendableInvokeCallback.onDone();
-  }
-
-  // TODO(#8578): Replace finalizer with PhantomReference.
-  @SuppressWarnings("deprecation")
-  protected void finalize() throws Throwable {
-    super.finalize();
-
-    if (callbackHandle != 0) {
-      deleteCallback(callbackHandle);
-      callbackHandle = 0;
-    }
+    wrappedBatchInvokeCallback.onDone();
   }
 }

--- a/src/controller/java/src/chip/devicecontroller/BatchInvokeCallbackJni.java
+++ b/src/controller/java/src/chip/devicecontroller/BatchInvokeCallbackJni.java
@@ -34,12 +34,16 @@ public final class BatchInvokeCallbackJni {
     this.wrappedBatchInvokeCallback = wrappedBatchInvokeCallback;
     this.callbackHandle = newCallback();
 
-    this.cleanable = Cleaner.create().register(this, () -> {
-      if (chipClusterPtr != 0) {
-        deleteCluster(chipClusterPtr);
-        chipClusterPtr = 0;
-      }
-    });
+    this.cleanable =
+        Cleaner.create()
+            .register(
+                this,
+                () -> {
+                  if (chipClusterPtr != 0) {
+                    deleteCluster(chipClusterPtr);
+                    chipClusterPtr = 0;
+                  }
+                });
   }
 
   long getCallbackHandle() {

--- a/src/controller/java/src/chip/devicecontroller/ChipDeviceController.java
+++ b/src/controller/java/src/chip/devicecontroller/ChipDeviceController.java
@@ -26,6 +26,7 @@ import chip.devicecontroller.model.ChipAttributePath;
 import chip.devicecontroller.model.ChipEventPath;
 import chip.devicecontroller.model.DataVersionFilter;
 import chip.devicecontroller.model.InvokeElement;
+import java.lang.ref.Cleaner;
 import java.time.ZoneOffset;
 import java.util.ArrayList;
 import java.util.Calendar;
@@ -43,6 +44,8 @@ public class ChipDeviceController {
   private CompletionListener completionListener;
   private ScanNetworksListener scanNetworksListener;
   private NOCChainIssuer nocChainIssuer;
+
+  private final Cleaner.Cleanable cleanable;
 
   /**
    * To load class and jni, we need to new AndroidChipPlatform after jni load but before new
@@ -67,6 +70,13 @@ public class ChipDeviceController {
       throw new NullPointerException("params cannot be null");
     }
     deviceControllerPtr = newDeviceController(params);
+
+    this.cleanable = Cleaner.create().register(this, () -> {
+      if (deviceControllerPtr != 0) {
+        deleteDeviceController(deviceControllerPtr);
+        deviceControllerPtr = 0;
+      }
+    });
   }
 
   public void setCompletionListener(CompletionListener listener) {
@@ -1771,16 +1781,6 @@ public class ChipDeviceController {
 
   static {
     System.loadLibrary("CHIPController");
-  }
-
-  @SuppressWarnings("deprecation")
-  protected void finalize() throws Throwable {
-    super.finalize();
-
-    if (deviceControllerPtr != 0) {
-      deleteDeviceController(deviceControllerPtr);
-      deviceControllerPtr = 0;
-    }
   }
 
   /** Interface to implement custom operational credentials issuer (NOC chain generation). */

--- a/src/controller/java/src/chip/devicecontroller/ChipDeviceController.java
+++ b/src/controller/java/src/chip/devicecontroller/ChipDeviceController.java
@@ -71,12 +71,16 @@ public class ChipDeviceController {
     }
     deviceControllerPtr = newDeviceController(params);
 
-    this.cleanable = Cleaner.create().register(this, () -> {
-      if (deviceControllerPtr != 0) {
-        deleteDeviceController(deviceControllerPtr);
-        deviceControllerPtr = 0;
-      }
-    });
+    this.cleanable =
+        Cleaner.create()
+            .register(
+                this,
+                () -> {
+                  if (deviceControllerPtr != 0) {
+                    deleteDeviceController(deviceControllerPtr);
+                    deviceControllerPtr = 0;
+                  }
+                });
   }
 
   public void setCompletionListener(CompletionListener listener) {

--- a/src/controller/java/src/chip/devicecontroller/ExtendableInvokeCallbackJni.java
+++ b/src/controller/java/src/chip/devicecontroller/ExtendableInvokeCallbackJni.java
@@ -34,12 +34,16 @@ public final class ExtendableInvokeCallbackJni {
     this.wrappedExtendableInvokeCallback = wrappedExtendableInvokeCallback;
     this.callbackHandle = newCallback();
 
-    this.cleanable = Cleaner.create().register(this, () -> {
-      if (callbackHandle != 0) {
-        deleteCallback(callbackHandle);
-        callbackHandle = 0;
-      }
-    });
+    this.cleanable =
+        Cleaner.create()
+            .register(
+                this,
+                () -> {
+                  if (callbackHandle != 0) {
+                    deleteCallback(callbackHandle);
+                    callbackHandle = 0;
+                  }
+                });
   }
 
   long getCallbackHandle() {

--- a/src/controller/java/src/chip/devicecontroller/ExtendableInvokeCallbackJni.java
+++ b/src/controller/java/src/chip/devicecontroller/ExtendableInvokeCallbackJni.java
@@ -19,6 +19,7 @@ package chip.devicecontroller;
 
 import chip.devicecontroller.model.InvokeResponseData;
 import chip.devicecontroller.model.NoInvokeResponseData;
+import java.lang.ref.Cleaner;
 import java.util.Optional;
 import javax.annotation.Nullable;
 
@@ -27,9 +28,18 @@ public final class ExtendableInvokeCallbackJni {
   private final ExtendableInvokeCallback wrappedExtendableInvokeCallback;
   private long callbackHandle;
 
+  private final Cleaner.Cleanable cleanable;
+
   public ExtendableInvokeCallbackJni(ExtendableInvokeCallback wrappedExtendableInvokeCallback) {
     this.wrappedExtendableInvokeCallback = wrappedExtendableInvokeCallback;
     this.callbackHandle = newCallback();
+
+    this.cleanable = Cleaner.create().register(this, () -> {
+      if (callbackHandle != 0) {
+        deleteCallback(callbackHandle);
+        callbackHandle = 0;
+      }
+    });
   }
 
   long getCallbackHandle() {
@@ -79,16 +89,5 @@ public final class ExtendableInvokeCallbackJni {
 
   private void onDone() {
     wrappedExtendableInvokeCallback.onDone();
-  }
-
-  // TODO(#8578): Replace finalizer with PhantomReference.
-  @SuppressWarnings("deprecation")
-  protected void finalize() throws Throwable {
-    super.finalize();
-
-    if (callbackHandle != 0) {
-      deleteCallback(callbackHandle);
-      callbackHandle = 0;
-    }
   }
 }

--- a/src/controller/java/src/chip/devicecontroller/GetConnectedDeviceCallbackJni.java
+++ b/src/controller/java/src/chip/devicecontroller/GetConnectedDeviceCallbackJni.java
@@ -17,14 +17,25 @@
  */
 package chip.devicecontroller;
 
+import java.lang.ref.Cleaner;
+
 /** JNI wrapper callback class for getting a connected device. */
 public class GetConnectedDeviceCallbackJni {
   private final GetConnectedDeviceCallback wrappedCallback;
   private long callbackHandle;
 
+  private final Cleaner.Cleanable cleanable;
+
   public GetConnectedDeviceCallbackJni(GetConnectedDeviceCallback wrappedCallback) {
     this.wrappedCallback = wrappedCallback;
     this.callbackHandle = newCallback(wrappedCallback);
+
+    this.cleanable = Cleaner.create().register(this, () -> {
+      if (callbackHandle != 0) {
+        deleteCallback(callbackHandle);
+        callbackHandle = 0;
+      }
+    });
   }
 
   long getCallbackHandle() {
@@ -34,17 +45,6 @@ public class GetConnectedDeviceCallbackJni {
   private native long newCallback(GetConnectedDeviceCallback wrappedCallback);
 
   private native void deleteCallback(long callbackHandle);
-
-  // TODO(#8578): Replace finalizer with PhantomReference.
-  @SuppressWarnings("deprecation")
-  protected void finalize() throws Throwable {
-    super.finalize();
-
-    if (callbackHandle != 0) {
-      deleteCallback(callbackHandle);
-      callbackHandle = 0;
-    }
-  }
 
   /** Callbacks for getting a device connected with PASE or CASE, depending on the context. */
   public interface GetConnectedDeviceCallback {

--- a/src/controller/java/src/chip/devicecontroller/GetConnectedDeviceCallbackJni.java
+++ b/src/controller/java/src/chip/devicecontroller/GetConnectedDeviceCallbackJni.java
@@ -30,12 +30,16 @@ public class GetConnectedDeviceCallbackJni {
     this.wrappedCallback = wrappedCallback;
     this.callbackHandle = newCallback(wrappedCallback);
 
-    this.cleanable = Cleaner.create().register(this, () -> {
-      if (callbackHandle != 0) {
-        deleteCallback(callbackHandle);
-        callbackHandle = 0;
-      }
-    });
+    this.cleanable =
+        Cleaner.create()
+            .register(
+                this,
+                () -> {
+                  if (callbackHandle != 0) {
+                    deleteCallback(callbackHandle);
+                    callbackHandle = 0;
+                  }
+                });
   }
 
   long getCallbackHandle() {

--- a/src/controller/java/src/chip/devicecontroller/InvokeCallbackJni.java
+++ b/src/controller/java/src/chip/devicecontroller/InvokeCallbackJni.java
@@ -17,8 +17,8 @@
  */
 package chip.devicecontroller;
 
-import java.lang.ref.Cleaner;
 import chip.devicecontroller.model.InvokeElement;
+import java.lang.ref.Cleaner;
 
 /** JNI wrapper callback class for {@link InvokeCallback}. */
 public final class InvokeCallbackJni {
@@ -31,12 +31,16 @@ public final class InvokeCallbackJni {
     this.wrappedInvokeCallback = wrappedInvokeCallback;
     this.callbackHandle = newCallback();
 
-    this.cleanable = Cleaner.create().register(this, () -> {
-      if (callbackHandle != 0) {
-        deleteCallback(callbackHandle);
-        callbackHandle = 0;
-      }
-    });
+    this.cleanable =
+        Cleaner.create()
+            .register(
+                this,
+                () -> {
+                  if (callbackHandle != 0) {
+                    deleteCallback(callbackHandle);
+                    callbackHandle = 0;
+                  }
+                });
   }
 
   long getCallbackHandle() {

--- a/src/controller/java/src/chip/devicecontroller/InvokeCallbackJni.java
+++ b/src/controller/java/src/chip/devicecontroller/InvokeCallbackJni.java
@@ -17,6 +17,7 @@
  */
 package chip.devicecontroller;
 
+import java.lang.ref.Cleaner;
 import chip.devicecontroller.model.InvokeElement;
 
 /** JNI wrapper callback class for {@link InvokeCallback}. */
@@ -24,9 +25,18 @@ public final class InvokeCallbackJni {
   private final InvokeCallback wrappedInvokeCallback;
   private long callbackHandle;
 
+  private final Cleaner.Cleanable cleanable;
+
   public InvokeCallbackJni(InvokeCallback wrappedInvokeCallback) {
     this.wrappedInvokeCallback = wrappedInvokeCallback;
     this.callbackHandle = newCallback();
+
+    this.cleanable = Cleaner.create().register(this, () -> {
+      if (callbackHandle != 0) {
+        deleteCallback(callbackHandle);
+        callbackHandle = 0;
+      }
+    });
   }
 
   long getCallbackHandle() {
@@ -54,16 +64,5 @@ public final class InvokeCallbackJni {
 
   private void onDone() {
     wrappedInvokeCallback.onDone();
-  }
-
-  // TODO(#8578): Replace finalizer with PhantomReference.
-  @SuppressWarnings("deprecation")
-  protected void finalize() throws Throwable {
-    super.finalize();
-
-    if (callbackHandle != 0) {
-      deleteCallback(callbackHandle);
-      callbackHandle = 0;
-    }
   }
 }

--- a/src/controller/java/src/chip/devicecontroller/ReportCallbackJni.java
+++ b/src/controller/java/src/chip/devicecontroller/ReportCallbackJni.java
@@ -20,6 +20,7 @@ package chip.devicecontroller;
 import chip.devicecontroller.model.ChipAttributePath;
 import chip.devicecontroller.model.ChipEventPath;
 import chip.devicecontroller.model.NodeState;
+import java.lang.ref.Cleaner;
 import javax.annotation.Nullable;
 
 /** JNI wrapper callback class for {@link ReportCallback}. */
@@ -30,6 +31,8 @@ public class ReportCallbackJni {
   private long callbackHandle;
   @Nullable private NodeState nodeState;
 
+  private final Cleaner.Cleanable cleanable;
+
   public ReportCallbackJni(
       @Nullable SubscriptionEstablishedCallback subscriptionEstablishedCallback,
       ReportCallback reportCallback,
@@ -39,6 +42,13 @@ public class ReportCallbackJni {
     this.wrappedResubscriptionAttemptCallback = resubscriptionAttemptCallback;
     this.callbackHandle =
         newCallback(subscriptionEstablishedCallback, resubscriptionAttemptCallback);
+
+    this.cleanable = Cleaner.create().register(this, () -> {
+      if (callbackHandle != 0) {
+        deleteCallback(callbackHandle);
+        callbackHandle = 0;
+      }
+    });
   }
 
   long getCallbackHandle() {
@@ -87,16 +97,5 @@ public class ReportCallbackJni {
 
   private void onDone() {
     wrappedReportCallback.onDone();
-  }
-
-  // TODO(#8578): Replace finalizer with PhantomReference.
-  @SuppressWarnings("deprecation")
-  protected void finalize() throws Throwable {
-    super.finalize();
-
-    if (callbackHandle != 0) {
-      deleteCallback(callbackHandle);
-      callbackHandle = 0;
-    }
   }
 }

--- a/src/controller/java/src/chip/devicecontroller/ReportCallbackJni.java
+++ b/src/controller/java/src/chip/devicecontroller/ReportCallbackJni.java
@@ -43,12 +43,16 @@ public class ReportCallbackJni {
     this.callbackHandle =
         newCallback(subscriptionEstablishedCallback, resubscriptionAttemptCallback);
 
-    this.cleanable = Cleaner.create().register(this, () -> {
-      if (callbackHandle != 0) {
-        deleteCallback(callbackHandle);
-        callbackHandle = 0;
-      }
-    });
+    this.cleanable =
+        Cleaner.create()
+            .register(
+                this,
+                () -> {
+                  if (callbackHandle != 0) {
+                    deleteCallback(callbackHandle);
+                    callbackHandle = 0;
+                  }
+                });
   }
 
   long getCallbackHandle() {

--- a/src/controller/java/src/chip/devicecontroller/WriteAttributesCallbackJni.java
+++ b/src/controller/java/src/chip/devicecontroller/WriteAttributesCallbackJni.java
@@ -19,6 +19,7 @@ package chip.devicecontroller;
 
 import chip.devicecontroller.model.ChipAttributePath;
 import chip.devicecontroller.model.Status;
+import java.lang.ref.Cleaner;
 import javax.annotation.Nullable;
 
 /** JNI wrapper callback class for {@link WriteAttributesCallback}. */
@@ -26,9 +27,18 @@ public final class WriteAttributesCallbackJni {
   private final WriteAttributesCallback wrappedWriteAttributesCallback;
   private long callbackHandle;
 
+  private final Cleaner.Cleanable cleanable;
+
   public WriteAttributesCallbackJni(WriteAttributesCallback wrappedWriteAttributesCallback) {
     this.wrappedWriteAttributesCallback = wrappedWriteAttributesCallback;
     this.callbackHandle = newCallback();
+
+    this.cleanable = Cleaner.create().register(this, () -> {
+      if (callbackHandle != 0) {
+        deleteCallback(callbackHandle);
+        callbackHandle = 0;
+      }
+    });
   }
 
   long getCallbackHandle() {
@@ -60,16 +70,5 @@ public final class WriteAttributesCallbackJni {
 
   private void onDone() {
     wrappedWriteAttributesCallback.onDone();
-  }
-
-  // TODO(#8578): Replace finalizer with PhantomReference.
-  @SuppressWarnings("deprecation")
-  protected void finalize() throws Throwable {
-    super.finalize();
-
-    if (callbackHandle != 0) {
-      deleteCallback(callbackHandle);
-      callbackHandle = 0;
-    }
   }
 }

--- a/src/controller/java/src/chip/devicecontroller/WriteAttributesCallbackJni.java
+++ b/src/controller/java/src/chip/devicecontroller/WriteAttributesCallbackJni.java
@@ -33,12 +33,16 @@ public final class WriteAttributesCallbackJni {
     this.wrappedWriteAttributesCallback = wrappedWriteAttributesCallback;
     this.callbackHandle = newCallback();
 
-    this.cleanable = Cleaner.create().register(this, () -> {
-      if (callbackHandle != 0) {
-        deleteCallback(callbackHandle);
-        callbackHandle = 0;
-      }
-    });
+    this.cleanable =
+        Cleaner.create()
+            .register(
+                this,
+                () -> {
+                  if (callbackHandle != 0) {
+                    deleteCallback(callbackHandle);
+                    callbackHandle = 0;
+                  }
+                });
   }
 
   long getCallbackHandle() {

